// koa-parser/lib/index.js
const parser = require('./parser')

module.exports = (opts = {
  encoding: 'utf-8', // 编码
  json: true, // 支持json解析
  text: true, // 支持text解析
  urlencoded: true, // 支持urlencoded解析
  multipart: true, // 支持form-data解析
  error: false // 解析错误回调
}) => {
  return async (ctx, next) => {
    if (ctx.request.body !== undefined) {
      return await next()
    }

    // 默认请求body
    ctx.request.body = {}
    try {
      if (opts.json && ctx.is('json')) { // 解析json
        ctx.request.body = await parser.json(ctx, {
          encoding: opts.encoding
        })
      } else if (opts.urlencoded && ctx.is('urlencoded')) { // 解析urlencoded
        ctx.request.body = await parser.urlencoded(ctx, {
          encoding: opts.encoding
        })
      } else if (opts.text && ctx.is('text')) { // 解析text
        ctx.request.body = await parser.text(ctx, {
          encoding: opts.encoding
        })
      }
      if (opts.multipart && ctx.is('multipart')) { // 解析multipart
        ctx.request.body = await parser.multipart(ctx, {
          encoding: opts.encoding
        })
      }
    } catch (err) {
      if (typeof opts.error === 'function') {
        opts.error(err)
      } else {
        throw err
      }
    }
    return await next()
  }
}
